### PR TITLE
[MIG] business_requirement: Migration to v12

### PR DESCRIPTION
**Pull request by martintorresvm**
cc @Tecnativa

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please install locally first before proposing here, as the module doesn't install.

**Comment by martintorresvm** [comment]
@pedrobaeza sorry but that. Indeed I had to change and test quite before push force again

**Comment by martintorresvm** [comment]
@Tardo @pedrobaeza please review

File: business_requirement/models/business.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+    @api.multi
+    def write(self, vals):
+        if vals.get('state'):
+            br_xml_id = self.env.ref(
+                'business_requirement.group_business_requirement_manager')
+            user = self.env['res.users']
+            grps = self.env.user.groups_id.ids
+            date = fields.Datetime.now()
+            if vals['state'] == 'confirmed':
+                vals.update({'confirmed_user_id': user,
```
Comment (Tardo) [inline business_requirement/models/business.py]: 'user' is an model object, not a record.

File: business_requirement/models/business.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+                             'confirmation_date': date})
+            if vals['state'] == 'draft':
+                vals.update({'confirmed_user_id': False,
+                             'approved_id': False,
+                             'confirmation_date': False,
+                             'approval_date': False
+                             })
+            if vals['state'] == 'approved':
+                if br_xml_id.id in grps:
+                    vals.update({'approved_id': user,
```
Comment (Tardo) [inline business_requirement/models/business.py]: 'user' is an model object, not a record.

File: business_requirement/models/business.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+            if vals['state'] in ('approved', 'in_progress',
+                                 'done'):
+                if br_xml_id.id not in grps:
+                    raise ValidationError(_(
+                        'You can only move to the following stage: '
+                        'draft/confirmed/cancel/drop.'))
+        return super(BusinessRequirement, self).write(vals)
+
+    @api.model
+    def _selection_get_states(self):
```
Comment (Tardo) [inline business_requirement/models/business.py]: I don't understand why need this having 'selection_add' on fields.Selection?

**Review by Tardo — CHANGES_REQUESTED** [review]

File: business_requirement/models/business.py (hunk 3 above)
Comment (martintorresvm) [reply business_requirement/models/business.py]: @Tardo I never used this. I tried and is not working. I checked where is used on others examples and always for inherited models so it's to add states as the name suggest.

The point of have a function is split states definition from field definition but no necessary to do it for sure

**Comment by martintorresvm** [comment]
Changes done @Tardo

File: business_requirement/__init__.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# © 2017-2019 Elico Corp (https://www.elico-corp.com).
```
Comment (Tardo) [inline business_requirement/__init__.py]:
```suggestion
# Copyright 2017-2019 Elico Corp (https://www.elico-corp.com).
```

File: business_requirement/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# © 2017-2019 Elico Corp (https://www.elico-corp.com).
```
Comment (Tardo) [inline business_requirement/__manifest__.py]:
```suggestion
# Copyright 2017-2019 Elico Corp (https://www.elico-corp.com).
```

File: business_requirement/models/__init__.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# © 2017 Elico Corp (https://www.elico-corp.com).
```
Comment (Tardo) [inline business_requirement/models/__init__.py]:
```suggestion
# Copyright 2017 Elico Corp (https://www.elico-corp.com).
```

File: business_requirement/models/business.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+                    raise ValidationError(_(
+                        'You can only move to the following stage: '
+                        'draft/confirmed /cancel/drop.'))
+            if vals['state'] in ('approved', 'in_progress',
+                                 'done'):
+                if br_xml_id.id not in grps:
+                    raise ValidationError(_(
+                        'You can only move to the following stage: '
+                        'draft/confirmed/cancel/drop.'))
+        return super(BusinessRequirement, self).write(vals)
```
Comment (Tardo) [inline business_requirement/models/business.py]:
```suggestion
        return super().write(vals)
```

File: business_requirement/tests/__init__.py [hunk 8]
```diff
@@ -0,0 +1,3 @@
+# © 2017-2019 Elico Corp (https://www.elico-corp.com).
```
Comment (Tardo) [inline business_requirement/tests/__init__.py]:
```suggestion
# Copyright 2017-2019 Elico Corp (https://www.elico-corp.com).
```

File: business_requirement/tests/test_br.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
+# © 2017-2019 Elico Corp (https://www.elico-corp.com).
```
Comment (Tardo) [inline business_requirement/tests/test_br.py]:
```suggestion
# Copyright 2017-2019 Elico Corp (https://www.elico-corp.com).
```

File: business_requirement/controllers/portal.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+from odoo.tools import consteq
+from odoo.addons.portal.controllers.portal import (
+    CustomerPortal,
+    pager as portal_pager, get_records_pager)
+
+
+class CustomerPortal(CustomerPortal):
+
+    def _prepare_portal_layout_values(self):
+        values = super(CustomerPortal, self)._prepare_portal_layout_values()
```
Comment (Tardo) [inline business_requirement/controllers/portal.py]:
```suggestion
        values = super()._prepare_portal_layout_values()
```

File: business_requirement/models/business.py [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+            'date', fields.Date.context_today(self)),
+        required=True
+    )
+
+    @api.model
+    def create(self, vals):
+        if vals.get('name', '/') == '/':
+            vals['name'] = self.env['ir.sequence'].next_by_code(
+                'business.requirement')
+        return super(BusinessRequirement, self).create(vals)
```
Comment (Tardo) [inline business_requirement/models/business.py]:
```suggestion
        return super().create(vals)
```

File: business_requirement/models/business.py [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+        for br in self:
+            formatted_name = u'[{}] {}'.format(br.name, br.description)
+            result.append((br.id, formatted_name))
+        return result
+
+    @api.model
+    def name_search(self, name, args=None, operator='ilike', limit=100):
+        """Search BR based on Name or Description"""
+        # Make a search with default criteria
+        names = super(BusinessRequirement, self).name_search(
```
Comment (Tardo) [inline business_requirement/models/business.py]:
```suggestion
        names = super().name_search(
```

File: business_requirement/models/business.py [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+            states = self.env['business.requirement'].\
+                fields_get(['state']).get('state').get('selection')
+            read_group_all_states = [{'__context': {'group_by': groupby[1:]},
+                                      '__domain': domain + [('state', '=',
+                                                             state_value)],
+                                      'state': state_value,
+                                      'state_count': 0}
+                                     for state_value, state_name in states]
+            # Get standard results
+            read_group_res = super(BusinessRequirement, self).\
```
Comment (Tardo) [inline business_requirement/models/business.py]:
```suggestion
            read_group_res = super().\
```

File: business_requirement/models/business.py [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+                        read_group_res))
+                if not res:
+                    res = list(
+                        filter(
+                            lambda x: x['state'] == state_value,
+                            read_group_all_states))
+                res[0]['state'] = [state_value, state_name]
+                result.append(res[0])
+            return result
+        return super(BusinessRequirement, self).\
```
Comment (Tardo) [inline business_requirement/models/business.py]:
```suggestion
        return super().\
```

File: business_requirement/models/business.py [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+            read_group(domain, fields, groupby,
+                       offset=offset, limit=limit, orderby=orderby, lazy=lazy)
+
+    def get_portal_confirmation_action(self):
+        return self.env['ir.config_parameter'].sudo().get_param(
+            'business_requirement.br_portal_confirmation_options',
+            default='none')
+
+    def _compute_portal_url(self):
+        super(BusinessRequirement, self)._compute_portal_url()
```
Comment (Tardo) [inline business_requirement/models/business.py]:
```suggestion
        super()._compute_portal_url()
```

**Review by Tardo — CHANGES_REQUESTED** [review]

**Comment by martintorresvm** [comment]
Changes done @Tardo

**Comment by martintorresvm** [comment]
@pedrobaeza Could you do a final review in order to merge?

File: business_requirement/__manifest__.py [hunk 16]
```diff
@@ -1,11 +1,11 @@
...
 {
     "name": "Business Requirement",
     "category": "Business Requirements Management",
     "summary": "Manage the Business Requirements (stories, scenarios, gaps\
         and test cases) for your customers",
-    "version": "11.0.1.0.0",
+    "version": "12.0.1.0.0",
     "website": "https://github.com/OCA/business-requirement",
     "author": "Elico Corp, "
               "Odoo Community Association (OCA)",
```
Comment (pedrobaeza) [inline business_requirement/__manifest__.py:12]: Where is Tecnativa here?!

File: business_requirement/data/business_requirement_demo.xml [hunk 17]
```diff
@@ -21,7 +21,7 @@
 
         <record id="business_requirement_1" model="business.requirement">
             <field name="description">Corporate Website</field>
-            <field name="responsible_user_id" ref="base.user_root"/>
+            <field name="responsible_user_id" ref="base.demo_user0"/>
```
Comment (pedrobaeza) [inline business_requirement/data/business_requirement_demo.xml]: Use `base.user_admin`, not a portal user.

File: business_requirement/data/business_requirement_demo.xml [hunk 18]
```diff
@@ -130,7 +130,7 @@ pulvinar et, cursus sed purus.
             <field name="category_ids"
                    eval="[(6, 0, [ref('business_requirement_category_2')])]"/>
             <field name="reviewer_ids"
-                   eval="[(6, 0, [ref('base.user_root')])]"/>
+                   eval="[(6, 0, [ref('base.demo_user0')])]"/>
```
Comment (pedrobaeza) [inline business_requirement/data/business_requirement_demo.xml]: The same

File: business_requirement/data/business_requirement_demo.xml [hunk 19]
```diff
@@ -142,7 +142,7 @@ pulvinar et, cursus sed purus.
...
```
Comment (pedrobaeza) [inline business_requirement/data/business_requirement_demo.xml]: The same for all the rest of coincidences

File: business_requirement/models/business.py (hunk 3 above)
Comment (pedrobaeza) [reply business_requirement/models/business.py]: Please remove this method and put directly `selection=[('draft', 'Draft'), ...]`

File: business_requirement/models/business.py [hunk 20]
```diff
@@ -265,10 +265,8 @@ def name_search(self, name, args=None, operator='ilike', limit=100):
         return list(set(names) | set(descriptions))[:limit]
 
     @api.multi
-    @api.returns('self', lambda value: value.id)
-    def message_post(self, body='', subject=None, message_type='notification',
-                     subtype=None, parent_id=False, attachments=None,
-                     content_subtype='html', **kwargs):
+    @api.returns('mail.message', lambda value: value.id)
+    def message_post(self, **kwargs):
```
Comment (pedrobaeza) [inline business_requirement/models/business.py]: Why not putting full signature if you are later putting specific keyword arguments?

File: business_requirement/security/security.xml [hunk 21]
```diff
@@ -6,7 +6,7 @@
         <field name="implied_ids" eval="[(4, ref('base.group_user'))]"/>
         <field name="category_id"
                ref="business_requirement.module_category_business_requirement"/>
-        <field name="users" eval="[(4, ref('base.user_root'))]"/>
+        <field name="users" eval="[(4, ref('base.user_admin'))]"/>
```
Comment (pedrobaeza) [inline business_requirement/security/security.xml:9]: This is indeed the correct user. Why not doing in the other place as well?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by martintorresvm** [comment]
Changes done @pedrobaeza

File: business_requirement/models/business.py [hunk 22]
```diff
@@ -0,0 +1,348 @@
...
+                'business.requirement')
+        return super().create(vals)
+
+    @api.multi
+    def write(self, vals):
+        if vals.get('state'):
+            br_xml_id = self.env.ref(
+                'business_requirement.group_business_requirement_manager')
+            user = self.env.user
+            grps = self.env.user.groups_id.ids
```
Comment (Tardo) [inline business_requirement/models/business.py]: In this case .ids order is irrelevant and 'set' it's faster to determining if an element is present. I see that "grps" is used only for searches purposes.
```suggestion
            grps = set(self.env.user.groups_id.ids)
```

File: business_requirement/models/business.py [hunk 23]
```diff
@@ -0,0 +1,348 @@
...
+                             })
+            if vals['state'] == 'approved':
+                if br_xml_id.id in grps:
+                    vals.update({'approved_id': user.id,
+                                 'approval_date': date})
+                else:
+                    raise ValidationError(_(
+                        'You can only move to the following stage: '
+                        'draft/confirmed /cancel/drop.'))
+            if vals['state'] in ('approved', 'in_progress',
```
Comment (Tardo) [inline business_requirement/models/business.py]:
```suggestion
            if vals['state'] in {'approved', 'in_progress',
```

File: business_requirement/models/business.py [hunk 24]
```diff
@@ -0,0 +1,348 @@
...
+                                 'done'):
```
Comment (Tardo) [inline business_requirement/models/business.py]:
```suggestion
                                 'done'}:
```

**Review by Tardo — CHANGES_REQUESTED** [review]

File: business_requirement/models/business.py (hunk 22 above)
Comment (pedrobaeza) [reply business_requirement/models/business.py]: No need. There are no duplicates in that field. ORM ensures that.

File: business_requirement/models/business.py (hunk 24 above)
Comment (pedrobaeza) [reply business_requirement/models/business.py]: I suppose this is to make a set (so the opening one should be changed also) for improving the `in` operator efficiency. Isn't it?

Comment (Tardo) [reply business_requirement/models/business.py]: Yes. The previous comment it's for opening.

File: business_requirement/models/business.py (hunk 22 above)
Comment (Tardo) [reply business_requirement/models/business.py]: First comment updated

Comment (pedrobaeza) [reply business_requirement/models/business.py]: Well, I would use `user.has_group` method instead, but this set optimization is not very decisive for the performance

**Comment by martintorresvm** [comment]
Changes done @Tardo @pedrobaeza

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**